keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+Note that the generated code for built-in will look like this:
+```golang
+func GenerateDefault(obj *Object) {
+     if obj.Name == "" {
+          obj.Name = "default-name"
+     }
+     if obj.Number == 0 {
+          obj.Number = 0
+     }
+}
+```
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users. This
+can be mitigated by double-checking the existing tests and adding new
+tests for gap in coverage.
+
+## Design Details
+
+We are proposing to modify the defaulting generator to also generate new
+defaulting functions, which will use the marker to generate the default
+and apply them to the fields directly if the fields are set to their
+zero-value. These new defaulting functions will be called, inside the
+existing defaulting logic, right before the existing defaulting
+functions. The defaulting call will stay unchanged.
+
+Because we also want server-side apply to be aware of these defaults,
+the OpenAPI generator will also use this information to fill the
+defaults fields in the OpenAPI. Then, Server-side apply will be able to use
+that additional information to default associative keys as needed.
+
+In other words, the plan would be the following:
+1. Update defaults generator to parse the `+default` marker and generate
+the additional defaulting functions, automatically call them prior to
+the existing defaulting functions
+2. Update kube-openapi to parse the `+default` marker and include it in
+the OpenAPI definition.
+3. Implement the new CRD semantics that clears `null` values from
+objects when the `x-kubernetes-remove-nulls` flag is present and true at
+the top-level (or in the CRD).
+4. Add additional "default" field to sigs.k8s.io/structured-merge-diff
+internal schema, and parse that field from the OpenAPI, use it when the
+map key is missing. The code for this already exists in a
+[PR](https://github.com/kubernetes-sigs/structured-merge-diff/pull/83#issuecomment-498410353)
+5. Update some of the existing built-in APIs to specify their default
+when they are strictly declarative, probably focusing first on defaults
+that are part of the associative keys (e.g. `protocol: TCP`).
+
+Note that this is not changing how built-in defaults work, but is merely
+adding an extra-layer on top of that which makes type authors life easier.
+
+### Test Plan
+
+As soon as this is implemented and working, we can start removing a lot
+of the existing defaulting code and replace it with this values, and we
+can re-use the existing defaulting tests, make sure that the behavior
+doesn't change when moving from code-based defaults to declarative
+default. Also panic in specific places of the defaulters to guarantee
+that the defaulting have happened before.
+
+### Graduation Criteria
+
+As an internal feature, this will go straight to stable.
```

Comment:
I think the graduation criteria should be that all eligible manually-written default functions are converted to this mechanism and the manual code deleted.  That would leave only "weird" cases in */defaults.go.

Reply:
if dynamic defaulting leaks into user manually-written defaults, it might be a bit more complex to replicate this in static markers. @apelisse did you discuss this case in the KEP / comments; i may have missed it?

```
func myDefaultingFunc(obj *SomeObject) {
	if obj.foo == 0 {
		obj.foo = 2
	}
	obj.bar = obj.foo - 1
}
```

Reply:
Right, that's why the "declarative" marker defaults will be applied after custom defaulting logic.

In your example, if there is a declarative marker for `foo` or `bar`, neither will be applied since they both have 2 and (that defaulting function is weird since `bar` is always set ;-) 1 respectively.

Reply:
Also, clearly, declarative markers won't be able to have this kind of logic, that's on purpose, more often than not, you're doing something wrong.

Reply:
> Right, that's why the "declarative" marker defaults will be applied after custom defaulting logic.

interesting, i would have expected the dynamic defaulting to apply after the static defaulting from markers.

> In your example, if there is a declarative marker for foo or bar, neither will be applied since they both have 2 and (that defaulting function is weird since bar is always set ;-) 1 set respectively.

i guess, i've tried to showcase that `bar`'s defaulting may solely depend on the value of `foo`.

Reply:
> interesting, i would have expected the dynamic (custom) defaulting to apply after the static defaulting from markers.

I haven't been able to think of an example where it would be better one way or another, but this is possibly simpler to reason about. What's left unspecified after the custom function will have the declarative value applied.

> i guess, i've tried to showcase that bar's defaulting may solely depend on the value of foo.

Yes, that's what I understood. But we're specifically not allowing this kind of cross-field defaulting

Reply:
> Also, clearly, declarative markers won't be able to have this kind of logic, that's on purpose, more often than not, you're doing something wrong if you need this.

hm, if so, there has to be a way to alert authors of new APIs about similar bad practices (if considered bad practices).

Reply:
Good point, I think we should.